src/utils/data-utils.ts

```diff
@@ -217,6 +217,49 @@ interface ColumnDataShape {
   units?: string;
 }
 
+interface LevelMap {
+  [key: string]: { description: string; standardizedTerm: string };
+}
+
+function buildCategoricalLevels(
+  allValues: string[] | undefined,
+  missingValues: string[] | undefined,
+  dictLevels: { [key: string]: { Description: string; TermURL?: string } } | undefined,
+  annotationLevels: { [key: string]: { TermURL: string; Label: string } } | undefined,
+  standardizedTerms: StandardizedTerms
+): LevelMap {
+  const missingValuesSet = new Set(missingValues ?? []);
+  const initialLevels = Array.from(new Set(allValues ?? []))
+    .filter((value) => !missingValuesSet.has(value))
+    .reduce(
+      (acc, value) => ({
+        ...acc,
+        [value]: { description: '', standardizedTerm: '' },
+      }),
+      {} as LevelMap
+    );
+
+  const resolvedDictLevels = dictLevels ?? {};
+  const resolvedAnnotationLevels = annotationLevels ?? {};
```

Comment:
I still find this a bit confusing. I.e. what is the benefit over just doing `dictLevels?.[value]` below? You're doing `dictLevels -> resolvedDictLevels (what is that?) -> dictLevel`. I think the reason you're doing this is to guard against `dictLevels` being `undefined` - but this implementation creates a bit of clutter